src/telegram/process_update_task.rs

```diff
@@ -39,24 +43,66 @@ impl ProcessUpdateTask {
         log::info!("Received a message {:?}", self.update);
 
         let api = AsyncApi::new(&RUST_TELEGRAM_BOT_TOKEN);
+        let db_controller = DbController::new().await?;
 
         if let UpdateContent::Message(message) = &self.update.content {
             let (chat_id, user_id, user) = Self::get_info_from_message(message);
 
             Self::send_typing(message, &api).await?;
 
-            let state = Self::fetch_state(&chat_id, user_id, user.clone()).await?;
+            let state = Self::fetch_state(&db_controller, &chat_id, user_id, user.clone()).await?;
 
             match state {
                 ClientState::Initial => match message.text.as_deref() {
                     Some("/start") | Some("/start@RustWeather77Bot") => {
                         Self::start(message, &user, &api).await?;
                     }
+
+                    Some("/pattern") | Some("/pattern@RustWeather77Bot") => {
+                        Self::pattern_city(message, &user, &api).await?;
+                        db_controller
+                            .modify_state(&chat_id, user_id, ClientState::FindCity)
+                            .await?;
+                    }
+
                     _ => {}
                 },
 
-                ClientState::Pattern => match message.text.as_deref() {
-                    Some("/pattern") | Some("/pattern@RustWeather77Bot") => {}
+                ClientState::FindCity => match message.text.as_deref() {
+                    Some("/cancel") | Some("/cancel@RustWeather77Bot") => {
+                        Self::cancel(&db_controller, message, &api).await?;
+                    }
+                    Some(text) => {
+                        if (Self::find_city(&db_controller, &user, message, &api).await).is_ok() {
+                            db_controller
+                                .modify_selected(&chat_id, user_id, text.to_string())
+                                .await?;
+                            db_controller
+                                .modify_state(&chat_id, user_id, ClientState::Number)
+                                .await?;
+                        }
+                    }
+
+                    _ => {}
+                },
+                ClientState::Number => match message.text.as_deref() {
+                    Some("/cancel") | Some("/cancel@RustWeather77Bot") => {
+                        Self::cancel(&db_controller, message, &api).await?;
+                    }
+                    Some(text) => match text.parse::<usize>() {
+                        Ok(_) => {
+                            Self::pattern_response(&db_controller, message, &api).await?;
+                            db_controller
```

Comment:
Every time when you modify the state, you change two fields. what do you think about creating a function that modifies two fields at the same time?

Reply:
[cancel logic out , improve command management , return to initial func](https://github.com/pxp9/weather_bot_rust/pull/9/commits/66ee70e20d78f7b22cca58b8de083f17486de2e9)